skiko/src/nativeJsMain/cpp/TextLine.cc

```diff
@@ -166,89 +98,75 @@ SKIKO_EXPORT void org_jetbrains_skia_TextLine__1nGetGlyphs
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
+    return instance->fRuns.size();
 }
-#endif
-
 
+SKIKO_EXPORT void org_jetbrains_skia_TextLine__1nGetRunPositions
+  (KNativePointer ptr, KFloat* resultArray) {
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
+    size_t size = instance->fRuns.size();
 
-SKIKO_EXPORT KFloat* org_jetbrains_skia_TextLine__1nGetBreakPositions
-  (KNativePointer ptr) {
-    TODO("implement org_jetbrains_skia_TextLine__1nGetBreakPositions");
+    for (size_t idx = 0; idx < size; ++idx)
+        resultArray[idx] = instance->fRuns[idx].fPosition;
 }
 
-#if 0
-SKIKO_EXPORT KFloat* org_jetbrains_skia_TextLine__1nGetBreakPositions
+SKIKO_EXPORT KInt org_jetbrains_skia_TextLine__1nGetBreakPositionsCount
   (KNativePointer ptr) {
-    TextLine* instance = reinterpret_cast<TextLine*>((ptr));
-    std::vector<KFloat> positions;
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
+    size_t count = 0;
     for (auto& run: instance->fRuns)
-        positions.insert(positions.end(), run.fBreakPositions.begin(), run.fBreakPositions.end());
-    return javaFloatArray(env, positions);
+        count += run.fBreakPositions.size();
+    return count;
 }
-#endif
-
 
-
-SKIKO_EXPORT KInt* org_jetbrains_skia_TextLine__1nGetBreakOffsets
-  (KNativePointer ptr) {
-    TODO("implement org_jetbrains_skia_TextLine__1nGetBreakOffsets");
+SKIKO_EXPORT void org_jetbrains_skia_TextLine__1nGetBreakPositions
+  (KNativePointer ptr, KFloat* resultArray) {
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
+    size_t added = 0;
+    for (auto& run: instance->fRuns) {
+        size_t count = run.fBreakPositions.size();
+        std::memcpy(resultArray + added, run.fBreakPositions.data(), count * sizeof(SkScalar));
+        added += count;
+    }
 }
 
-#if 0
-SKIKO_EXPORT KInt* org_jetbrains_skia_TextLine__1nGetBreakOffsets
+
+SKIKO_EXPORT KInt org_jetbrains_skia_TextLine__1nGetBreakOffsetsCount
   (KNativePointer ptr) {
-    TextLine* instance = reinterpret_cast<TextLine*>((ptr));
-    std::vector<KInt> offsets;
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
+    size_t count = 0;
     for (auto& run: instance->fRuns)
-        offsets.insert(offsets.end(), run.fBreakOffsets.begin(), run.fBreakOffsets.end());
-    return javaIntArray(env, offsets);
+        count += run.fBreakOffsets.size();
+    return count;
 }
-#endif
-
 
+SKIKO_EXPORT void org_jetbrains_skia_TextLine__1nGetBreakOffsets
+  (KNativePointer ptr, KInt* resultArray) {
+    TextLine* instance = reinterpret_cast<TextLine*>(ptr);
 
-SKIKO_EXPORT KInt org_jetbrains_skia_TextLine__1nGetOffsetAtCoord
-  (KNativePointer ptr, KFloat x) {
-    TODO("implement org_jetbrains_skia_TextLine__1nGetOffsetAtCoord");
+    size_t added = 0;
+    for (auto& run: instance->fRuns) {
+        size_t count = run.fBreakOffsets.size();
+        std::memcpy(resultArray + added, run.fBreakOffsets.data(), count * sizeof(uint32_t));
```

Comment:
may be we should refer in such places to the size of the array we actualy expect to get (in out case sizeof(KInt)?

Reply:
done